### PR TITLE
JIT Layernorm fusion

### PR DESCRIPTION
**Pull request by wanchaol**
Partially fuse layer_norm by decomposing layer_norm into the batchnorm kernel that computes the stats, and then fusing the affine operations after the reduce operations, this is similar to the batchnorm fusion that @apaszke did, it also only works in inference mode now.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -12322,6 +12322,50 @@ def forward(self, x, y):
...
+        model_noopt.load_state_dict(model.state_dict())
+        x = torch.randn(2, 10, 5, 5, device='cuda')
+        y = torch.randn(2, 10, 5, 5, device='cuda')
+        # FIXME: We need differentiation for CNNs for this optimization to trigger
+        with torch.no_grad():
+            out = model(x, y)
+            graph = model.graph_for(x, y)
+            rep = str(graph)
+
+            print(rep)
```
Comment (zou3519) [inline test/test_jit.py]: Is this print supposed to be here?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 2]
```diff
@@ -494,6 +570,15 @@ struct GraphFuser {
     if (consumer->kind() != prim::FusionGroup) {
       group = createSingletonFusionGroup(consumer);
     }
+    if (producer->node()->matches(
+            "aten::layer_norm(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True) -> Tensor")) {
+      // We don't do any fusions in here, but simply decompose the batch norm
```
Comment (zou3519) [inline torch/csrc/jit/passes/graph_fuser.cpp]: s/batch norm/layer norm/

File: test/test_jit.py [hunk 3]
```diff
@@ -12322,6 +12322,50 @@ def forward(self, x, y):
         fused_graph = fusion_groups[0].g('Subgraph')
         self.assertTrue(any(node.kind() == 'aten::sqrt' for node in fused_graph.nodes()))
 
+    @unittest.skipIf(IS_WINDOWS, "NYI: fuser support for Windows")
+    @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA")
+    @skipIfRocm
+    def test_fuse_layer_norm(self):
```
[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_jit.py]: Yes please! This test is huge

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -244,6 +275,58 @@ struct GraphFuser {
...
+    auto bias_defined = isDefined(bias).value();
+    if (weight_defined && bias_defined) {
+      new_output = graph_->insert(aten::addcmul, {bias, new_output, weight});
+    } else if (weight_defined) {
+      new_output = graph_->insert(aten::mul, {new_output, weight});
+    } else if (bias_defined) {
+      new_output = graph_->insert(aten::add, {new_output, bias});
+    }
+    layer_norm->output()->replaceAllUsesWith(new_output);
+    layer_norm->destroy();
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: This infra should also be shared. THey take almost the same arguments, only need a different reshape and replacement.

**Comment by wanchaol** [comment]
@apaszke @ngimel I merged batchnorm and layernorm infrastructure together. Please take a look again when you get some time

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 5]
```diff
@@ -244,6 +266,167 @@ struct GraphFuser {
...
+        new_output = graph_->insert(aten::mul, {new_output, weight});
+      } else if (bias_defined) {
+        new_output = graph_->insert(aten::add, {new_output, bias});
+      }
+      normalization_op->output()->replaceAllUsesWith(new_output);
+      normalization_op->destroy();
+    }
+  }
+
+  void decomposeLayerNorm(Node* layer_norm) {
```
Comment (ngimel) [inline torch/csrc/jit/passes/graph_fuser.cpp]: you don't need this anymore

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 6]
```diff
@@ -244,6 +266,167 @@ struct GraphFuser {
...
+    } else if (weight_defined) {
+      new_output = graph_->insert(aten::mul, {new_output, weight});
+    } else if (bias_defined) {
+      new_output = graph_->insert(aten::add, {new_output, bias});
+    }
+    layer_norm->output()->replaceAllUsesWith(new_output);
+    layer_norm->destroy();
+  }
+
   void decomposeBatchNorm(Node* batch_norm) {
```
Comment (ngimel) [inline torch/csrc/jit/passes/graph_fuser.cpp]: you don't need this anymore, you have common function

**Review by ngimel — CHANGES_REQUESTED** [review]
Please remove code that is no longer used.

**Review by ngimel — APPROVED** [review]
